test/testdata/infer/private_constant.rb

```diff
@@ -0,0 +1,55 @@
+# typed: true
+
+module Foo
+  A_PUBLIC_CONST = 'public'
+  A_PRIVATE_CONST = 'private'
+  private_constant :A_PRIVATE_CONST
+
+  ANOTHER_PRIVATE_CONST = true
+  private_constant 'ANOTHER_PRIVATE_CONST'
+
+  class PrivateClass; end
+  private_constant :PrivateClass
+
+  class AnotherPrivateClass; end
+  private_constant 'AnotherPrivateClass'
+
+  module PrivateModule
+    def self.ok_private_usage
+      A_PUBLIC_CONST
+      A_PRIVATE_CONST
+      PrivateClass
+      PrivateModule
+    end
+
+    class ClassInsidePrivateModule
+      def self.also_ok_private_usage; end
+    end
+  end
+  private_constant :PrivateModule
+
+  module AnotherPrivateModule; end
+  private_constant 'AnotherPrivateModule'
+
+  def self.ok_private_usage
+    A_PUBLIC_CONST
+    A_PRIVATE_CONST
+    PrivateClass
+    PrivateModule.ok_private_usage
+    PrivateModule::ClassInsidePrivateModule.also_ok_private_usage
+  end
+
+  def self.not_ok_private_usage
+    ::Foo::A_PUBLIC_CONST
+    ::Foo::A_PRIVATE_CONST # error: private constant Foo::A_PRIVATE_CONST referenced
+    ::Foo::PrivateClass # error: private constant Foo::PrivateClass referenced
+    ::Foo::PrivateModule # error: private constant Foo::PrivateModule referenced
+  end
+end
```

Comment:
Can you add tests for

- private type aliases (type aliases are `static-field` symbols that have a special bit set on them to make them type aliases, so they should work without further modification but I'd like to test
- usages of private classes / modules / type aliases / class aliases in signatures

this is a class alias:

```ruby
# typed: true
A = Integer
private_constant :A

extend T::Sig
sig {params(x: A).void}
def foo(x); end
```

Can you also add a test for type member symbols?

```ruby
# typed: true
extend T::Sig

class A
  extend T::Generic
  Elem = type_member

  class B
    extend T::Sig
    sig {params(x: Elem).void}
    def bar(x); end
  end
end

sig {params(x: A::Elem).void}
def foo(x); end
```
[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20A%0A%20%20extend%20T%3A%3AGeneric%0A%20%20Elem%20%3D%20type_member%0A%0A%20%20class%20B%0A%20%20%20%20extend%20T%3A%3ASig%0A%20%20%20%20sig%20%7Bparams(x%3A%20Elem).void%7D%0A%20%20%20%20def%20bar(x)%3B%20end%0A%20%20end%0Aend%0A%0Asig%20%7Bparams(x%3A%20A%3A%3AElem).void%7D%0Adef%20foo(x)%3B%20end)

We already have some ad-hoc logic to make sure that you don't refer to type members outside of their class, because the rules that we want to implement are actually stricter than `private_constant` (see the example above—`bar`'s sig has a private reference to the constant, but that's not valid because `Elem` wasn't defined in `B`, but in `A`).

Given that we must keep the special handling of type members, I'd suggest that you implement straightforward support for private type members the same way as static fields. Even though our checks for proper type member usage are stricter than private checks, i don't think we attempt to build those type member checks on top of private checks, because those constants will not actually be marked private at runtime.